<h1 align="center">Introduction to Machine Learning</h1>
<h4 align="center">Instructor : Dr. Shamsollahi</h4>
<h4 align="center">Electerical Engineering department of Sharif University of Technology, Spring 2024</h4>
<h4 align="center">Computer Assignment 2</h4>
<h4 align="center">Arman Yazdani 400102255</h4>

## Diagnosis of cardiovascular diseases
Cardiovascular diseases are the most important cause of death of Iranians, diseases which, according to the statistics of the Ministry of Health and Civil Registry, take the lives of at least 140,000 people every year and make thousands of people hospitalized, disabled and disabled. Diseases that depend more on lifestyle than any other factors such as nutrition, movement and exercise, smoking and air pollution and can be prevented to a large extent by controlling these factors. Our goal in this exercise is to use information 
People's health standard and different classification methods in machine learning to predict the existence of this category of diseases in people. The collection of data at your disposal 
It contains the information of 2000 patients. There are 18 columns in this data set, which are as follows from left to right:$\\$
Column Names: ['age', 'gender', 'height', 'weight', 'ap_hi', 'ap_lo', 'smoke', 'alco', 'active', 'cardio','cholesterol_1', 'cholesterol_2', 'cholesterol_3', 'gluc_1', 'gluc_2', 'gluc_3', 'bmi', 'bp']$\newline$
The variable 'cardio' is actually our label, which determines whether a person is sick (1 = cardio) or not (0 = cardio). In view of these cases 
Answer the following questions:

$\large{1.\hspace{0.2cm} Data-set\quad preparation}\newline$
1.1:
Read the dataset first and display the first five rows. Also, with the help of the $\textit{.info()}$ command, display different information about the data set.

In [19]:
import pandas as pd
# Read CSV file
data = pd.read_csv("cardio_pred.csv")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            1997 non-null   float64
 1   gender         1997 non-null   float64
 2   height         1991 non-null   float64
 3   weight         1995 non-null   float64
 4   ap_hi          1996 non-null   float64
 5   ap_lo          1996 non-null   float64
 6   smoke          1998 non-null   float64
 7   alco           1995 non-null   float64
 8   active         1996 non-null   float64
 9   cardio         1996 non-null   float64
 10  cholesterol_1  1994 non-null   float64
 11  cholesterol_2  1991 non-null   float64
 12  cholesterol_3  1996 non-null   float64
 13  gluc_1         1993 non-null   float64
 14  gluc_2         1996 non-null   float64
 15  gluc_3         1993 non-null   float64
 16  bmi            1997 non-null   float64
 17  bp             1993 non-null   float64
dtypes: float

In [12]:
print(data.iloc[:, :5])

       age  gender  height  weight  ap_hi
0     51.0     1.0   168.0    62.0  110.0
1     56.0     0.0   156.0    85.0  140.0
2     52.0     0.0   165.0    64.0  130.0
3     49.0     1.0   169.0    82.0  150.0
4     49.0     0.0     NaN    56.0  100.0
...    ...     ...     ...     ...    ...
1995  51.0     0.0   160.0    58.0  120.0
1996  40.0     1.0   176.0    75.0  160.0
1997  46.0     0.0   164.0    74.0  120.0
1998  55.0     1.0   164.0    75.0  120.0
1999  63.0     0.0   155.0    72.0  160.0

[2000 rows x 5 columns]


1.2: In this dataset, some of the cells are empty (missing values), remove the corresponding rows from this dataset. Then repeat the previous section.$\\$
$\textit{\small{Check the indices on the left hand side}}$

In [13]:
data1 = data.dropna()
print(data1.iloc[:, :5])

       age  gender  height  weight  ap_hi
0     51.0     1.0   168.0    62.0  110.0
6     62.0     0.0   157.0    93.0  130.0
7     63.0     1.0   178.0    95.0  130.0
12    41.0     1.0   165.0    60.0  120.0
14    40.0     1.0   181.0    95.0  130.0
...    ...     ...     ...     ...    ...
1995  51.0     0.0   160.0    58.0  120.0
1996  40.0     1.0   176.0    75.0  160.0
1997  46.0     0.0   164.0    74.0  120.0
1998  55.0     1.0   164.0    75.0  120.0
1999  63.0     0.0   155.0    72.0  160.0

[1937 rows x 5 columns]


In [14]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1937 entries, 0 to 1999
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            1937 non-null   float64
 1   gender         1937 non-null   float64
 2   height         1937 non-null   float64
 3   weight         1937 non-null   float64
 4   ap_hi          1937 non-null   float64
 5   ap_lo          1937 non-null   float64
 6   smoke          1937 non-null   float64
 7   alco           1937 non-null   float64
 8   active         1937 non-null   float64
 9   cardio         1937 non-null   float64
 10  cholesterol_1  1937 non-null   float64
 11  cholesterol_2  1937 non-null   float64
 12  cholesterol_3  1937 non-null   float64
 13  gluc_1         1937 non-null   float64
 14  gluc_2         1937 non-null   float64
 15  gluc_3         1937 non-null   float64
 16  bmi            1937 non-null   float64
 17  bp             1937 non-null   float64
dtypes: float64(18

1.3: Divide the data into a random 80% for train and 20% for test results.(set random state=2024)

In [54]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Dropping cardio
X = data1.drop('cardio',axis=1)
Y = data1['cardio']
# Normalaizing (Z-Score) & Getting DataFrame
scaler = StandardScaler()
X_norm = scaler.fit_transform(X)
X_DataFrame = pd.DataFrame(X_norm, columns = X.columns)
# Splitting Data
X_train, X_test, Y_train, Y_test = train_test_split(X_DataFrame,Y,test_size=0.2,random_state=2024)

$\large{2.\hspace{0.2cm}Logistic\hspace{0.2cm}Regression\hspace{0.2cm}Classifier}$

2.1:
By training a logistic regression model, classify the data and get the accuracy of the model.

In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Perform Logistic Regression
model = LogisticRegression()
model.fit(X_train, Y_train)

# Evaluate the model
y_pred = model.predict(X_test)
accuracy = accuracy_score(Y_test, y_pred)

print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.75


2.2 Now we want to identify the most important features. Calculate the coefficients of the model defined in the previous section and explain the relationship between each feature and the output. Which features are more important and why؟

In [ ]:
# Get the coefficients of the model
coefficients = model.coef_[0]

# Print the coefficients with feature names
feature_names = X.columns
# Create a list of tuples (feature_name, coefficient)
coef_tuples = list(zip(feature_names, coefficients))

# Sort the list of tuples by the absolute value of the coefficient
coef_tuples.sort(key=lambda x: abs(x[1]), reverse=True)

# Print the sorted coefficients
for feature, coef in coef_tuples:
    print(f"Feature: {feature}, Coefficient: {coef:.2f}")


Feature: ap_hi, Coefficient: 0.90
Feature: age, Coefficient: 0.42
Feature: cholesterol_3, Coefficient: 0.24
Feature: height, Coefficient: 0.21
Feature: bmi, Coefficient: 0.16
Feature: ap_lo, Coefficient: 0.14
Feature: gender, Coefficient: -0.14
Feature: cholesterol_1, Coefficient: -0.14
Feature: bp, Coefficient: 0.08
Feature: gluc_2, Coefficient: 0.07
Feature: gluc_3, Coefficient: -0.06
Feature: cholesterol_2, Coefficient: -0.06
Feature: smoke, Coefficient: -0.05
Feature: active, Coefficient: -0.04
Feature: alco, Coefficient: -0.03
Feature: weight, Coefficient: -0.01
Feature: gluc_1, Coefficient: -0.01


$\bullet$ The sign of the coefficient (positive or negative) indicates the direction of the relationship between the feature and the target variable. A positive coefficient means that as the feature value increases, the likelihood of the target variable being 1 (or the positive class) increases. A negative coefficient means that as the feature value increases, the likelihood of the target variable being 1 decreases.$\\$
$\bullet$ The magnitude of the coefficient represents the strength of the relationship between the feature and the target variable. Features with larger absolute coefficients have a stronger impact on the prediction of the target variable.

In the Cardiovascular Disease Prediction dataset, the feature ap_hi represents the systolic blood pressure of individuals. Systolic blood pressure is the higher value measured during a heartbeat and reflects the pressure in the arteries when the heart contracts.$\\$
Elevated systolic blood pressure is a major risk factor for cardiovascular diseases which makes it the most important factor of Cardiovascular Disease Prediction.

$\large{3.Support\hspace{0.1cm}Vector\hspace{0.1cm}Machine\hspace{0.1cm}Classifier}$

3.1 This time، we are going to classify the dataset with SVM. First, define your model using the linear kernel and calculate the accuracy of different C-values.

In [29]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
# Define the range of C values to test
C_values = [0.1, 1, 10, 100]

# Create a dictionary of parameters for GridSearchCV
param_grid = {'C': C_values}

# Create the SVM model with a linear kernel
svm_model = SVC(kernel='linear')

# Use GridSearchCV to find the best C value
grid_search = GridSearchCV(svm_model, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, Y_train)

# Print the results for all C values
print("Results for different C values:")
for c in C_values:
    svm_model = SVC(kernel='linear', C=c)
    svm_model.fit(X_train, Y_train)
    y_pred = svm_model.predict(X_test)
    test_accuracy = accuracy_score(Y_test, y_pred)
    print(f"C = {c}, Test accuracy: {test_accuracy:.2f}")

Results for different C values:
C = 0.1, Test accuracy: 0.74
C = 1, Test accuracy: 0.74
C = 10, Test accuracy: 0.75
C = 100, Test accuracy: 0.75


3.2 Now use the poly kernel and calculate the accuracy of the classifier from different values of C and degree parameters.

In [39]:
# Define the range of degree values to test
degree_values = [2, 3, 4]

# Create a dictionary of parameters for GridSearchCV
param_grid = {'C': C_values, 'degree': degree_values}

# Create the SVM model with a polynomial kernel
svm_model = SVC(kernel='poly')

# Use GridSearchCV to find the best C and degree values
grid_search = GridSearchCV(svm_model, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, Y_train)

# Print the results for all C and degree values
print("Results for different C and degree values:")
for c in C_values:
    for d in degree_values:
        svm_model = SVC(kernel='poly', C=c, degree=d)
        svm_model.fit(X_train, Y_train)
        y_pred = svm_model.predict(X_test)
        test_accuracy = accuracy_score(Y_test, y_pred)
        print(f"C = {c}, Degree = {d}, Test accuracy: {test_accuracy:.2f}")

Results for different C and degree values:
C = 0.1, Degree = 2, Test accuracy: 0.57
C = 0.1, Degree = 3, Test accuracy: 0.68
C = 0.1, Degree = 4, Test accuracy: 0.61
C = 1, Degree = 2, Test accuracy: 0.60
C = 1, Degree = 3, Test accuracy: 0.70
C = 1, Degree = 4, Test accuracy: 0.66
C = 10, Degree = 2, Test accuracy: 0.61
C = 10, Degree = 3, Test accuracy: 0.70
C = 10, Degree = 4, Test accuracy: 0.68
C = 100, Degree = 2, Test accuracy: 0.61
C = 100, Degree = 3, Test accuracy: 0.69
C = 100, Degree = 4, Test accuracy: 0.64


3.3 Finally, use the rbf kernel and calculate the accuracy of the classifier from different values of C and gamma parameters.

In [38]:
# Define the range of gamma values to test
gamma_values = [0.01, 0.1, 1, 10]

# Create a dictionary of parameters for GridSearchCV
param_grid = {'C': C_values, 'gamma': gamma_values}

# Create the SVM model with an RBF kernel
svm_model = SVC(kernel='rbf')

# Use GridSearchCV to find the best C and gamma values
grid_search = GridSearchCV(svm_model, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, Y_train)

# Print the results for all C and gamma values
print("Results for different C and gamma values:")
for c in C_values:
    for g in gamma_values:
        svm_model = SVC(kernel='rbf', C=c, gamma=g)
        svm_model.fit(X_train, Y_train)
        y_pred = svm_model.predict(X_test)
        test_accuracy = accuracy_score(Y_test, y_pred)
        print(f"C = {c}, Gamma = {g}, Test accuracy: {test_accuracy:.2f}")

Results for different C and gamma values:
C = 0.1, Gamma = 0.01, Test accuracy: 0.74
C = 0.1, Gamma = 0.1, Test accuracy: 0.71
C = 0.1, Gamma = 1, Test accuracy: 0.52
C = 0.1, Gamma = 10, Test accuracy: 0.51
C = 1, Gamma = 0.01, Test accuracy: 0.74
C = 1, Gamma = 0.1, Test accuracy: 0.71
C = 1, Gamma = 1, Test accuracy: 0.64
C = 1, Gamma = 10, Test accuracy: 0.52
C = 10, Gamma = 0.01, Test accuracy: 0.73
C = 10, Gamma = 0.1, Test accuracy: 0.70
C = 10, Gamma = 1, Test accuracy: 0.61
C = 10, Gamma = 10, Test accuracy: 0.54
C = 100, Gamma = 0.01, Test accuracy: 0.72
C = 100, Gamma = 0.1, Test accuracy: 0.67
C = 100, Gamma = 1, Test accuracy: 0.60
C = 100, Gamma = 10, Test accuracy: 0.54


3.4 Compare the accuracy calculated in the previous sections. Which kernel performed best with what parameters؟ What is the reason?

As you can see on part 2.2,the majority of the coefficients are small but one.which means training will result as an almost linear function of the biggest coefficient.
later on,we observed that $\\$
linear's best result was: $\\$
C=10,100 => ACC=75% $\\$
while Poly's was: $\\$
C=0.1,degree=3 => ACC=70% $\\$
and RBF's was: $\\$
C=0.1,gamma=0.01 => ACC=74% 

$\large{4.Perceptron\hspace{0.1cm}Classifier}$

4.1 This time, with the help of perceptron algorithm, classify the data and get it's correctness.Do you use the sample mode algorithm or batch mode, why? (You are not allowed to use the Perceptron function in this section.)

In [20]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load the data
data = pd.read_csv('cardio_pred.csv')
data = data.dropna()
# Dropping cardio
X = data.drop('cardio',axis=1)
y = data['cardio']
# Normalaizing (Z-Score) & Getting DataFrame
scaler = StandardScaler()
X_norm = scaler.fit_transform(X)
X_DataFrame = pd.DataFrame(X_norm, columns = X.columns)
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2024)

# Scale the data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Implement the Perceptron algorithm in batch mode
def perceptron_batch(X, y, epochs=100, learning_rate=0.1):
    n_samples, n_features = X.shape
    weights = np.zeros(n_features)
    bias = 0

    for _ in range(epochs):
        mistakes = 0
        for x, target in zip(X, y):
            prediction = np.dot(x, weights) + bias
            if np.sign(prediction) != target:
                weights += learning_rate * target * x
                bias += learning_rate * target
                mistakes += 1
        if mistakes == 0:
            break

    return weights, bias

# Train the Perceptron model in batch mode
weights, bias = perceptron_batch(X_train_scaled, y_train)

# Evaluate the model on the test set
y_pred = np.sign(np.dot(X_test_scaled, weights) + bias)
accuracy = np.mean(y_pred == y_test)
print(f"Test accuracy (batch mode): {accuracy:.2f}")

Test accuracy (batch mode): 0.49


In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Hyperparameters
input_size = 17  # Number of input features
hidden_size = 10  # Number of neurons in the hidden layer
output_size = 1  # Binary classification output
learning_rate = 0.01
num_epochs = 170
batch_size = 1

# Load the data
data = pd.read_csv('cardio_pred.csv')

# Data Cleaning
data = data.dropna()  # Drop rows with missing values

# Features and Labels
X = data.drop(columns=['cardio']).values
y = data['cardio'].values.reshape(-1, 1)

# Standardize the features
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Activation Functions
def relu(x):
    return np.maximum(0, x)

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Derivatives
def relu_derivative(x):
    return (x > 0).astype(float)

def sigmoid_derivative(x):
    s = sigmoid(x)
    return s * (1 - s)

# Initialize Weights and Biases
np.random.seed(42)
W1 = np.random.randn(input_size, hidden_size) * 0.01
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size) * 0.01
b2 = np.zeros((1, output_size))

# Forward and Backward Propagation
def forward_backward_propagation(X, y):
    global W1, b1, W2, b2

    # Forward pass
    Z1 = np.dot(X, W1) + b1
    A1 = relu(Z1)
    Z2 = np.dot(A1, W2) + b2
    A2 = sigmoid(Z2)

    # Compute loss (binary cross-entropy)
    m = y.shape[0]
    loss = -np.sum(y * np.log(A2) + (1 - y) * np.log(1 - A2)) / m

    # Backward pass
    dZ2 = A2 - y
    dW2 = np.dot(A1.T, dZ2) / m
    db2 = np.sum(dZ2, axis=0, keepdims=True) / m

    dA1 = np.dot(dZ2, W2.T)
    dZ1 = dA1 * relu_derivative(Z1)
    dW1 = np.dot(X.T, dZ1) / m
    db1 = np.sum(dZ1, axis=0, keepdims=True) / m

    # Update parameters
    W1 -= learning_rate * dW1
    b1 -= learning_rate * db1
    W2 -= learning_rate * dW2
    b2 -= learning_rate * db2

    return loss

# Training the model
for epoch in range(num_epochs):
    indices = np.random.permutation(X_train.shape[0])
    X_train_shuffled = X_train[indices]
    y_train_shuffled = y_train[indices]

    for i in range(0, X_train.shape[0], batch_size):
        X_batch = X_train_shuffled[i:i+batch_size]
        y_batch = y_train_shuffled[i:i+batch_size]

        loss = forward_backward_propagation(X_batch, y_batch)

    if epoch % 10 == 0:
        print(f'Epoch {epoch}, Loss: {loss}')

# Model Evaluation
def predict(X):
    Z1 = np.dot(X, W1) + b1
    A1 = relu(Z1)
    Z2 = np.dot(A1, W2) + b2
    A2 = sigmoid(Z2)
    return (A2 > 0.5).astype(int)

y_pred_train = predict(X_train)
y_pred_test = predict(X_test)

train_accuracy = np.mean(y_pred_train == y_train) * 100
test_accuracy = np.mean(y_pred_test == y_test) * 100

print(f'Training Accuracy: {train_accuracy:.2f}%')
print(f'Test Accuracy: {test_accuracy:.2f}%')


Epoch 0, Loss: 0.5319244320135192
Epoch 10, Loss: 0.1459605316894754
Epoch 20, Loss: 0.317531674232084
Epoch 30, Loss: 0.7981796544303684
Epoch 40, Loss: 0.7629133422618124
Epoch 50, Loss: 0.027593709582227857
Epoch 60, Loss: 0.8927954537183166
Epoch 70, Loss: 0.3364692997746001
Epoch 80, Loss: 0.28944126602478476
Epoch 90, Loss: 0.16608337290247377
Epoch 100, Loss: 0.12433663918163967
Epoch 110, Loss: 1.08808737945291
Epoch 120, Loss: 0.3918185050231273
Epoch 130, Loss: 0.1625598822075858
Epoch 140, Loss: 0.013261025802569956
Epoch 150, Loss: 0.13405288520744293
Epoch 160, Loss: 0.8041216217303436
Training Accuracy: 76.31%
Test Accuracy: 75.00%


It seem that since we had total package of data,It's a better idea to train with batch mode but the results didn't do so.this could be due to outlayer data or skipping a regional extermum.

4.2 This time, with the help of the Perceptron Built-in Function classification and compare its accuracy with the previous section.

In [4]:
import numpy as np
import pandas as pd
from sklearn.linear_model import Perceptron
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.preprocessing import StandardScaler

# Load the data
data = pd.read_csv('cardio_pred.csv')
data = data.dropna()
# Dropping cardio
X = data.drop('cardio',axis=1)
y = data['cardio']
# Normalaizing (Z-Score) & Getting DataFrame
scaler = StandardScaler()
X_norm = scaler.fit_transform(X)
X_DataFrame = pd.DataFrame(X_norm, columns = X.columns)
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2024)

# Scale the data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Use the built-in Perceptron function
perceptron = Perceptron(random_state=42)
perceptron.fit(X_train_scaled, y_train)

# Evaluate the model on the test set
y_pred = perceptron.predict(X_test_scaled)
accuracy_built_in = np.mean(y_pred == y_test)
print(f"Test accuracy (built-in Perceptron): {accuracy_built_in:.2f}")

Test accuracy (built-in Perceptron): 0.64


$\large{5.MLP\hspace{0.1cm}Neural\hspace{0.1cm} Networks}$

5.1 We are going to classify the data using neural networks (MLP). Research about different parameters of MLPClassifier (number of hidden neurons, activation functions, solver, etc.) and define your neural network using GridSearchCV and classify the data and report its accuracy.

In [5]:
from sklearn.neural_network import MLPClassifier
# Define the parameter grid
param_grid = {
    'hidden_layer_sizes': [(50,), (100,), (50, 50), (100, 100)],
    'activation': ['relu', 'tanh', 'logistic'],
    'solver': ['adam', 'sgd'],
    'alpha': [0.0001, 0.001, 0.01],
    'learning_rate': ['constant', 'adaptive']
}
# Create the MLPClassifier
mlp = MLPClassifier(random_state=42)

# Perform GridSearchCV
grid_search = GridSearchCV(mlp, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train_scaled, y_train)

# Print the best hyperparameters and the corresponding test accuracy
print(f"Best hyperparameters: {grid_search.best_params_}")
print(f"Test accuracy: {grid_search.score(X_test_scaled, y_test):.2f}")


Best hyperparameters: {'activation': 'logistic', 'alpha': 0.0001, 'hidden_layer_sizes': (100, 100), 'learning_rate': 'constant', 'solver': 'adam'}
Test accuracy: 0.73


c:\Users\HAMAHANG\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


This means that the best configuration found by the GridSearchCV is a neural network with two hidden layers, each with 100 neurons, using the logistic activation function, the Adam solver, a learning rate of 'constant', and an alpha (regularization) value of 0.0001

$\large{6.Decision\hspace{0.1cm}Tree}\\$

6.1 In this section، we are going to classify the data with the help of the decision tree. First، find the best parameters of the tree (such as tree depth، impurity criteria). And, then define a DecisionTreeClassifier random_state model with = 2024 and classify the data using the parameters you have obtained, and report its accuracy. (Hint: You can use GridSearchCV.)

In [11]:
from sklearn.tree import DecisionTreeClassifier
# Define the parameter grid
param_grid = {
    'max_depth': [3, 5, 7, 9, 11],
    'min_samples_split': [2, 4, 6, 8, 10],
    'criterion': ['gini', 'entropy']
}
# Create the DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state=2024)

# Perform GridSearchCV
grid_search = GridSearchCV(dt, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train_scaled, y_train)

# Print the best hyperparameters and the corresponding test accuracy
print(f"Best hyperparameters: {grid_search.best_params_}")
print(f"Test accuracy: {grid_search.score(X_test_scaled, y_test):.2f}")

Best hyperparameters: {'criterion': 'gini', 'max_depth': 5, 'min_samples_split': 2}
Test accuracy: 0.68


The resulting Accuracies were:$\\$
$\textit{Logistic Regression}$ : 75% $\\$
$\textit{SVM(Linear)}$ : 75% $\\$
$\textit{SVM(Polynomial)}$ : 70% $\\$
$\textit{SVM(RBF)}$ : 74% $\\$
$\textit{Manual Perceptron(BS=1)}$ : 75% $\\$
$\textit{Manual Perceptron(BS=dataSize)}$ : 49% $\\$
$\textit{Built-in Perceptron}$ : 64% $\\$
$\textit{MLP}$ : 73% $\\$
$\textit{Decision Tree}$ : 68% $\\$
where $\textit{Logistic Regression,SVM(Linear),Perceptron}$ performed better which might be due to the nature of data(it is well established that MLP specially with higher hidden layers work better),parameter of model and etc.$\\$
In addition to Acuracy we can estimate F1 Score,Percision,Least Squares Error and so on.